Update Payment Request button for Variable Products

asumaran opened the pull request:
# Changes proposed in this Pull Request:

Part of #1563. #1577 follow-up.

- Removes completely traces of `getSelectedProductData`/`get_selected_product_data `. Removal started in #1577. Now it's completely removed after confirming is not needed for variable products.
- Fixes an issue where Shipping amount was not correct after changing/updating Shipping Method.
- Fixes a validation issue that happened when a variation didn't have enough stock. Error is displayed correctly now.

# Testing instructions

This PR focuses on variable products only. Fortunately #1577 already covers many of the cases already so I focused here on cases that might broke for variable products only.

1. Create a "color" attribute.
2. Add terms "red", "green", "blue".
3. Create a variation from all attributes.
4. Apply different scenarios to it. You can use the same cases I've tested on the Test list below.
5. Check correct amount and item list is correct in Google Pay/Apple Pay.

#### Test list - Variable Product

- [x] Regular Price
- [x] Regular Price with default variation
- [x] With quantity
- [x] With Sale Price
- [x] Variation Without enough stock - Works better now with add-ons.
- [x] With Shipping
- [x] With Shipping, when changing Shipping Method
- [x] With Taxes
- [x] With Virtual products
- [x] Car Page With Coupon
- [x] With add-ons - This one actually works even better now
- [x] Cart Block
- [x] Checkout Block

asumaran committed:
Unlock button after alert when no variation is selected

asumaran committed:
Remove unnecessary event handler

asumaran committed:
Remove unused action and its handler

# Conflicts:
#	includes/payment-methods/class-wc-stripe-payment-request.php

# Conflicts:
#	includes/payment-methods/class-wc-stripe-payment-request.php

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1128,11 +1088,18 @@ public function ajax_add_to_cart() {
-				WC()->cart->add_to_cart( $product->get_id(), $qty );
+			if ( ! $has_enough_stock ) {
+				/* translators: 1: product name 2: quantity in stock */
+				throw new Exception( sprintf( __( 'You cannot add that amount of "%1$s"; to the cart because there is not enough stock (%2$s remaining).', 'woocommerce-gateway-stripe' ), $product->get_name(), wc_format_stock_quantity_for_display( $product->get_stock_quantity(), $product ) ) );
```
Shouldn't we throw this before attempting to add the product to the cart?

asumaran replied:
Nice catch. I'll update the code

ricardo reviewed (changes requested):
This tests well with variable products 💯 

I just left a comment regarding the stock check.

Shouldn't we merge this into `fix/1563-cart-total-for-prb`?

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1046,20 +1011,21 @@ public function ajax_add_to_cart() {
+			$variation_id     = 0;
+			$attributes       = [];
 
 			if ( ! is_a( $product, 'WC_Product' ) ) {
 				/* translators: %d is the product Id */
 				throw new Exception( sprintf( __( 'Product with the ID (%d) cannot be found.', 'woocommerce-gateway-stripe' ), $product_id ) );
 			}
```
We should probably add this check before trying to use any of `$product`'s properties/methods as in `$product->has_enough_stock( $qty );` and `$product->get_type();`.

asumaran replied:
Good catch! done a40fce9

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1073,13 +1039,19 @@ public function ajax_add_to_cart() {
+				throw new Exception( sprintf( __( 'You cannot add that amount of "%1$s"; to the cart because there is not enough stock (%2$s remaining).', 'woocommerce-gateway-stripe' ), $product->get_name(), wc_format_stock_quantity_for_display( $product->get_stock_quantity(), $product ) ) );
 			}
 
+			WC()->cart->add_to_cart( $product->get_id(), $qty, $variation_id, $attributes );
```
It looks fine to me to use the same `add_to_cart` for all these supported product types, but I wonder if by removing the checks for product type (`variable`, `variable-subscription`, `simple` and `subscription`), if we don't open up for any other issues.

Is `supported_product_types` enough to prevent issues?

asumaran replied:
I think it's fine. In any case we have a try/catch that will respond accordingly with a proper error message. I'll improve if it doesn't work well with subscriptions.

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1046,20 +1011,21 @@ public function ajax_add_to_cart() {
-			$qty        = ! isset( $_POST['qty'] ) ? 1 : absint( $_POST['qty'] );
+			$product_id       = isset( $_POST['product_id'] ) ? absint( $_POST['product_id'] ) : 0;
+			$product          = wc_get_product( $product_id );
+			$post_input       = filter_input_array( INPUT_POST, [ 'shipping_pending' => FILTER_VALIDATE_BOOLEAN ] );
+			$shipping_pending = $post_input['shipping_pending'];
```
Is `shipping_pending` a leftover?

asumaran replied:
Yes. it will be removed once I update the branch with the new changes from #1577

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1046,20 +1011,21 @@ public function ajax_add_to_cart() {
+			$product          = wc_get_product( $product_id );
+			$post_input       = filter_input_array( INPUT_POST, [ 'shipping_pending' => FILTER_VALIDATE_BOOLEAN ] );
+			$shipping_pending = $post_input['shipping_pending'];
+			$qty              = ! isset( $_POST['qty'] ) ? 1 : absint( $_POST['qty'] );
```
I know this isn't part of your code, so feel free to ignore this:

`qty` seems to be the only variable and posted property that is shortened. I guess we could rename it to `quantity`. It would be a tiny code readability improvement in my opinion.

We could rename the POST data passed from the front-end too.

asumaran replied:
done 1d16e16

ricardo reviewed (changes requested):
Code changes from previous review look good 👏 .

I had a question and a few minor comments.

asumaran committed:
Rename qty to quantity for readability

asumaran committed:
Throw error immediately if product doesn't exist

ricardo reviewed (approved):
LGTM 👍

asumaran committed:
Handle the case where variation has no enough stock

# Conflicts:
#	includes/payment-methods/class-wc-stripe-payment-request.php

# Conflicts:
#	includes/payment-methods/class-wc-stripe-payment-request.php
